LandSandBoat/server
# Fix item enhancement/effect: multiple effects, wear on zone. Add enhancement/effect source

**_I affirm:_**

## What does this pull request do?

Migrates a fix from ASB to LSB for Enhancements:

Originally adapted from PR by Tracent https://github.com/AirSkyBoat/AirSkyBoat/pull/2935:

- Allows multiple enchantment effects to be active on a player at the same time (was validated on retail by siknoz)
- Introduces the ability to add an effect by `EffectSourceType` and `EffectSourceTypeParam` which:
- Allows multiple enchantment effects to be active on a player at the same time
- Enchantment status effects now wear off when zoning
- Fixes an exploit where players could keep an enchantment even after unequipping the gear that gives the enchantment
- Adds status effects for items (listed below)
- Corrects status effects for items (listed below)
- Adds items to item.lua

## Steps to test these changes

- Use an item that grants an effect on use and check effect
- Use two effect items such as high breath mantle (15487) and vision ring (15559) to see two status effects and then remove gear one by one to see only correct enchantment removed
- Use aero mufflers then remove and see that enaero status effect remains
- Zone with an enchantment to see it wear off

First 60 seconds of this video is pre fix showing only 1 enhancement being applied at a time, not being removed when equipment is removed, and not removed on zone. 60s onwards is post fix from this pr.

https://www.dropbox.com/s/14sxrexj1vbfkdh/Post-Fix%20Enchantment.mp4?dl=0

Waiting on confirmations for a couple items but wanted to get an initial review/test the waters.

While this touches a lot of files, nearly all status effects are unchanged in terms of values - implementation is updated to use the `EffectSourceType` and `EffectSourceTypeParam` for tracking where an effect came from - this also sets up effects of other `EffectSourceType` such as `MOB`, `TEMPORARY_ITEM`, etc.

## Additions

| **Item**               |
|------------------------|
| Melt Baselard           |
| Aiming Gloves           |
| Death Chakram           |
| Hallowed Sword          |
| Hydra Tiara             |
| Hydra Tights            |
| Koen                   |
| Manashell Ring          |
| Memento Muffler         |
| Palmer's Bangles        |
| Prominence Axe          |
| Prominence Sword        |
| Purgatory Collar        |
| Sacred Degen            |
| Sacred Lance            |
| Sacred Mace             |
| Sacred Maul             |
| Sacred Sword            |
| Sacred Wand             |
| Talisman Obi            |
| Vulcan Blade            |
| Vulcan Claymore         |
| Vulcan Degen            |
| Vulcan Sword            |

---

## Changes

| **Item**           | **Change**                                                          |
|--------------------|---------------------------------------------------------------------|
| Pacifist Ring       | Corrected to use `ENMITY_DOWN`         |
| Tough Belt          | Corrected from `600s` to `60s`.                                    |

## Notes

| **Item**           | **Change**                                                          |
|--------------------|---------------------------------------------------------------------|
| Mighty Ring       | Should stack - need to track this by some instance based ID         |
